src/solver/multilabelsoftmarginloss/forward_reduced_multilabelsoftmarginloss.cpp

```diff
@@ -0,0 +1,230 @@
+#include <miopen/datatype.hpp>
+#include <miopen/kernel_build_params.hpp>
+#include <miopen/multilabelsoftmarginloss/invoke_params.hpp>
+#include <miopen/multilabelsoftmarginloss/solvers.hpp>
+#include <miopen/multilabelsoftmarginloss.hpp>
+#include <miopen/target_properties.hpp>
+#include <miopen/tensor_view_utils.hpp>
+
+#define LOCAL_SIZE_MULTILABELSOFTMARGINLOSS 256
+#define LOCAL_SIZE_REDUCE 256
+
+namespace miopen {
+
+namespace solver {
+
+namespace multilabelsoftmarginloss {
+
+bool MultilabelSoftMarginLossForward::IsApplicable(
+    const ExecutionContext& /*context*/,
+    const miopen::multilabelsoftmarginloss::ForwardProblemDescription& problem) const
+{
+    if(problem.GetiDesc().GetLengths()[1] > 24)
+        return false;
+    return true;
+}
+
+ConvSolution MultilabelSoftMarginLossForward::GetSolution(
+    const ExecutionContext& /*context*/,
+    const miopen::multilabelsoftmarginloss::ForwardProblemDescription& problem) const
+{
+    auto result = ConvSolution{miopenStatusSuccess};
+
+    auto xgrid = problem.GetiDesc().GetLengths()[0];
+
+    {
+        /* Phase 1: Calc loss for each element. */
+        auto dtype        = problem.GetiDesc().GetType();
+        size_t xlocalsize = LOCAL_SIZE_MULTILABELSOFTMARGINLOSS;
+        size_t xgridsize  = AlignUp(xgrid, xlocalsize);
+        size_t ylocalsize = 1;
+        size_t ygridsize  = 1;
+        size_t zlocalsize = 1;
+        size_t zgridsize  = 1;
+
+        auto kernel        = KernelInfo{};
+        kernel.kernel_file = "MIOpenMultilabelSoftMarginLoss.cpp";
+        kernel.kernel_name = "MultilabelSoftMarginLossForward2d";
+
+        const auto build_params = KernelBuildParameters{
+            {"MIOPEN_USE_FP16", static_cast<int32_t>(dtype == miopenHalf)},
+            {"MIOPEN_USE_FP32", static_cast<int32_t>(dtype == miopenFloat)},
+            {"MIOPEN_USE_BFP16", static_cast<int32_t>(dtype == miopenBFloat16)},
+            {"INPUT_TYPE",
+             (dtype == miopenBFloat16) ? "ushort"
+             : (dtype == miopenHalf)   ? "_Float16"
+                                       : "float"},
+        };
+
+        kernel.comp_options = build_params.GenerateFor(kbp::HIP{});
+
+        kernel.l_wk.push_back(xlocalsize);
+        kernel.l_wk.push_back(ylocalsize);
+        kernel.l_wk.push_back(zlocalsize);
+
+        kernel.g_wk.push_back(xgridsize);
+        kernel.g_wk.push_back(ygridsize);
+        kernel.g_wk.push_back(zgridsize);
+
+        result.construction_params.push_back(kernel);
+    }
+
+    {
+        /* Phase 2: Reduce */
+        auto _size              = xgrid;
+        auto dtype              = problem.GetiDesc().GetType();
+        const auto build_params = KernelBuildParameters{
+            {"MIOPEN_USE_FP16", static_cast<int32_t>(dtype == miopenHalf)},
+            {"MIOPEN_USE_FP32", static_cast<int32_t>(dtype == miopenFloat)},
+            {"MIOPEN_USE_BFP16", static_cast<int32_t>(dtype == miopenBFloat16)},
+            {"INPUT_TYPE",
```

Comment:
Same as above

Reply:
Fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/e9f959f5c1bfb65a1fb4033449671487b45c9bf0)